### PR TITLE
Feature/add testlists

### PR DESCRIPTION
**Pull request by lorenzoPrimi**

### CODE REVIEW

**Comment by willscott** [comment]
Do you have to do a new one of these each time the test lists get updated?

**Comment by lorenzoPrimi** [comment]
No, this is only for this release, in the future the test lists will be removed and the orchestrator will send me urls based on my location.

File: ooniprobe/Utility/NotificationService.mm [hunk 1]
```diff
@@ -86,13 +84,13 @@ - (void)registerNotifications{
...
-                                        error.as_ooni_error().c_str());
+                    //client.logger->warn("Register terminated with error: %s",
+                    //                    error.as_ooni_error().c_str());
                     return;
                 }
                 if (auth.dump(secrets_path) != mk::NoError()) {
-                    client.logger->warn("Cannot write secrets_path: %s",
-                                        error.as_ooni_error().c_str());
+                    //client.logger->warn("Cannot write secrets_path: %s",
+                    //                    error.as_ooni_error().c_str());
```
Comment (lorenzoPrimi) [inline ooniprobe/Utility/NotificationService.mm]: @bassosimone this

**Comment by bassosimone** [comment]
I see conflicts with master. I am gonna solve these.

**Comment by bassosimone** [comment]
Done!

**Comment by bassosimone** [comment]
(Maaaan, this PR looks huge!)

File: ooniprobe/Utility/TestLists.mm [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+@implementation TestLists
+
++ (id)sharedTestLists
+{
+    static TestLists *sharedTestLists = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedTestLists = [[self alloc] init];
+    });
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: Cannot you allocate when you declare?

File: ooniprobe/Utility/TestLists.mm [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        [self getCC];
+    }
+    return self;
+}
+
+- (NSArray*)getUrls{
+    NSMutableArray *urls = [[NSMutableArray alloc] init];
+    NSArray *global_urls = [self getUrlsForCountry:@"global"];
+    [urls addObjectsFromArray:global_urls];
+    NSArray *local_urls = [self getUrlsForCountry:[self.probe_cc lowercaseString]];
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: What if `probe_cc` is `nil`?

File: ooniprobe/Utility/TestLists.mm (hunk 2 above)
Comment (lorenzoPrimi) [reply ooniprobe/Utility/TestLists.mm]: It's something ARC related

File: ooniprobe/Utility/TestLists.mm (hunk 3 above)
Comment (bassosimone) [reply ooniprobe/Utility/TestLists.mm]: I also don't get exactly where is `probe_cc` initialized

File: ooniprobe/Utility/TestLists.mm [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    return nil;
+}
+
+- (void)getCC {
+    //TODO refactor all mk helper functions into a class
+    //This will be done with the next release as this class is going to be deleted
+    NSBundle *bundle = [NSBundle mainBundle];
+    NSString *geoip_asn_path = [bundle pathForResource:@"GeoIPASNum" ofType:@"dat"];
+    NSString *geoip_country_path = [bundle pathForResource:@"GeoIP" ofType:@"dat"];
+    mk::ooni::orchestrate::Client client;
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: I think here you may want to refactor the code to call the free function `find_location`.

File: ooniprobe/Utility/TestLists.mm (hunk 3 above)
Comment (lorenzoPrimi) [reply ooniprobe/Utility/TestLists.mm]: The next ```pathForResource:country_code ofType:@"csv"``` will return nil so local_urls will be an empty array

File: ooniprobe/Utility/TestLists.mm (hunk 4 above)
Comment (lorenzoPrimi) [reply ooniprobe/Utility/TestLists.mm]: done

File: scripts/update-test-lists.sh [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
```
Comment (bassosimone) [inline scripts/update-test-lists.sh]: You should split loooooooooong lines!

File: ooniprobe/View/TestInfoViewController.mm [hunk 6]
```diff
@@ -53,7 +53,9 @@ -(void)reloadTest{
 
 -(IBAction)runTest:(id)sender{
     NetworkMeasurement *current = [currentTests getTestWithName:testName];
-    current.inputs = nil;
+    //TODO change this code when refactor NetworkMeasurement class
+    [current setUri_scheme:NO];
```
Comment (bassosimone) [inline ooniprobe/View/TestInfoViewController.mm]: I am not sure I really like the name of this variable. Why is it named like this?

Comment (lorenzoPrimi) [reply ooniprobe/View/TestInfoViewController.mm]: uri_scheme is needed to disable max_runtime. do you have any name suggestion?

File: Podfile [hunk 7]
```diff
@@ -2,12 +2,13 @@ platform :ios, '9.0'
 target 'ooniprobe' do
     pod 'measurement_kit',
         :git => 'https://github.com/measurement-kit/measurement-kit.git',
-	:tag => 'v0.8.0-beta.1'
+	:tag => 'v0.8.0'
```
Comment (bassosimone) [inline Podfile]: You should update this 👎 <- /me is trolling here

File: ooniprobe/AppDelegate.mm [hunk 8]
```diff
@@ -139,6 +139,10 @@ - (void)applicationWillEnterForeground:(UIApplication *)application {
 
 - (void)applicationDidBecomeActive:(UIApplication *)application {
     // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
+    
+    //Updating the Country Code
+    [[TestLists sharedTestLists] getCC];
```
Comment (bassosimone) [inline ooniprobe/AppDelegate.mm]: @lorenzoPrimi can you rename this `updateCC`?

File: ooniprobe/Model/NetworkMeasurement.h [hunk 9]
```diff
@@ -29,6 +30,8 @@
 @property BOOL viewed;
 @property BOOL entry;
 @property int anomaly;
+@property BOOL uri_scheme;
```
Comment (bassosimone) [inline ooniprobe/Model/NetworkMeasurement.h]: @lorenzoPrimi can you rename this variable to have a more clear name?

File: ooniprobe/Model/NetworkMeasurement.mm [hunk 10]
```diff
@@ -230,21 +231,17 @@ -(void)run{
 }
 
 -(void) run_test {
-    NSBundle *bundle = [NSBundle mainBundle];
-    NSString *path = [bundle pathForResource:@"global" ofType:@"txt"];
     mk::nettests::WebConnectivityTest test;
     [super init_common:test];
+    //The bool uri_scheme to set the max_runtime, off when is true
```
Comment (bassosimone) [inline ooniprobe/Model/NetworkMeasurement.mm]: You can zap this comment

File: ooniprobe/Utility/TestLists.mm (hunk 2 above)
Comment (bassosimone) [reply ooniprobe/Utility/TestLists.mm]: ack

File: ooniprobe/Utility/TestLists.mm [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    static TestLists *sharedTestLists = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedTestLists = [[self alloc] init];
+    });
+    return sharedTestLists;
+}
+
+-(id)init
+{
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: It would be better to indent parentheses consistently

File: ooniprobe/Utility/TestLists.mm [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    NSArray *local_urls = [self getUrlsForCountry:[self.probe_cc lowercaseString]];
+    if (local_urls != nil)
+        [urls addObjectsFromArray:local_urls];
+    //NSLog(@"url %@", urls);
+    return urls;
+}
+
+- (NSArray*)getUrlsForCountry:(NSString*)country_code {
+    if ([[NSBundle mainBundle] pathForResource:country_code ofType:@"csv"] != nil){
+        NSURL *csvURL = [NSURL URLWithString:[NSString stringWithFormat:@"file://%@", [[NSBundle mainBundle] pathForResource:country_code ofType:@"csv"]]];
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: Can you please fold this very long line?

File: ooniprobe/Utility/TestLists.mm [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+    return nil;
+}
+
+- (void)updateCC {
+    //TODO refactor all mk helper functions into a class
+    //This will be done with the next release as this class is going to be deleted
+    NSBundle *bundle = [NSBundle mainBundle];
+    NSString *geoip_asn_path = [bundle pathForResource:@"GeoIPASNum" ofType:@"dat"];
+    NSString *geoip_country_path = [bundle pathForResource:@"GeoIP" ofType:@"dat"];
+    mk::ooni::find_location([geoip_country_path UTF8String], [geoip_asn_path UTF8String],{}, mk::Logger::global(), [self]
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: Can you fold this long line?

File: ooniprobe/Utility/TestLists.mm [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+- (NSArray*)getUrls {
+    NSMutableArray *urls = [[NSMutableArray alloc] init];
+    NSArray *global_urls = [self getUrlsForCountry:@"global"];
+    [urls addObjectsFromArray:global_urls];
+    NSArray *local_urls = [self getUrlsForCountry:[self.probe_cc lowercaseString]];
+    if (local_urls != nil)
+        [urls addObjectsFromArray:local_urls];
+    //NSLog(@"url %@", urls);
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: I can't merge with this comment, sorry

File: ooniprobe/Utility/TestLists.mm [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+@implementation TestLists
+
++ (id)sharedTestLists {
+    static TestLists *sharedTestLists = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedTestLists = [[self alloc] init];
+    });
+    return sharedTestLists;
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: Couldn't there be a race condition between initializing this singleton and using it? @lorenzoPrimi

Comment (bassosimone) [reply ooniprobe/Utility/TestLists.mm]: With @lorenzoPrimi we have checked that `dispatch_once` is "absolutely synchronous".

File: ooniprobe/Utility/TestLists.mm [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+-(id)init {
+    self = [super init];
+    if(self)
+    {
+        self.probe_cc = @"";
+        self.probe_asn = @"";
+        [self updateCC];
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: Can you rename this function such that it's clear that it is async?

File: ooniprobe/Utility/TestLists.mm [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    NSString *geoip_asn_path = [bundle pathForResource:@"GeoIPASNum" ofType:@"dat"];
+    NSString *geoip_country_path = [bundle pathForResource:@"GeoIP" ofType:@"dat"];
+    mk::ooni::find_location([geoip_country_path UTF8String],
+                            [geoip_asn_path UTF8String],{},
+                            mk::Logger::global(),
+                            [self]
+                            (mk::Error &&error, std::string probe_asn,
+                             std::string probe_cc) mutable {
+                                if (error) {
+                                    mk::warn("cannot find location");
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: Can you add here the error? The code should look like:

```C++
mk::warn("cannot find location: %s", error.what());
```

File: ooniprobe/Utility/TestLists.mm [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+                                    return;
+                                }
+                                self.probe_cc = [NSString stringWithFormat:@"%s", probe_cc.c_str()];
+                                self.probe_asn = [NSString stringWithFormat:@"%s", probe_asn.c_str()];
+                                mk::warn("probe_asn: %s", probe_asn.c_str());
+                                mk::warn("probe_cc: %s", probe_cc.c_str());
```
Comment (bassosimone) [inline ooniprobe/Utility/TestLists.mm]: You should not use `mk::warn` here. When there is no error, it's common to use `info`.

**Pull request merged**